pkg/xds/translate.go

```diff
@@ -40,8 +41,186 @@ const (
 	HeaderScheme    = ":scheme"
 )
 
-// ClusterName specifies cluster name for a destination
-type ClusterName func(*routingv2.Destination) string
+const (
+	// UnresolvedCluster for destinations pointing to unknown clusters.
+	UnresolvedCluster = "unresolved-cluster"
+
+	// DefaultOperation is the default decorator
+	DefaultOperation = "default-operation"
+)
+
+// ServiceByName claims a service entry from the registry using a host name.
+type ServiceByName func(host string, contextNamespace string) *model.Service
+
+// SubsetSelector resolves a subset to labels.
+type SubsetSelector func(service *model.Service, subset string) map[string]string
+
+// GuardedHost is a context-dependent virtual host entry with guarded routes.
+type GuardedHost struct {
+	// Port is the capture port (e.g. service port)
+	Port int
+
+	// Services are the services matching the virtual host.
+	// The service host names need to be contextualized by the source.
+	Services []*model.Service
+
+	// Hosts is a list of alternative literal host names for the host.
+	Hosts []string
+
+	// Routes in the virtual host
+	Routes []GuardedRoute
+}
+
+// TranslateVirtualHosts creates the entire routing table for Istio v1alpha2 configs.
+// Services are indexed by FQDN hostnames.
+// Cluster domain is used to resolve short service names (e.g. "svc.cluster.local").
+func TranslateVirtualHosts(
+	serviceConfigs []model.Config,
+	services map[string]*model.Service,
+	subsetSelector SubsetSelector,
+	clusterDomain string) []GuardedHost {
+	out := make([]GuardedHost, 0)
+
+	serviceByName := func(host string, contextNamespace string) *model.Service {
+		if strings.Contains(host, ".") {
+			return services[host]
+		}
+
+		return services[fmt.Sprintf("%s.%s.%s", host, contextNamespace, clusterDomain)]
```

Comment:
Bug: This is the kind of issue I was talking about on Slack. This is assuming that a DNS name is always made of three components, wherein namespace is part of the DNS name. This is applicable only to Kube. This same bug broke Consul few months ago.

You need more if conditions here.. if namespace is empty, skip the second dot..

Reply:
Can be done later? This is kept a function so that we can tweak it as we go along.

Reply:
sure